### PR TITLE
Added xattr plugin for darwin.

### PR DESCRIPTION
**Pull request by bmcdermott-r7**
Addressing issue #3273

### CODE REVIEW

File: vql/darwin/xattr.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+		arg := &XAttrArgs{}
+		err := arg_parser.ExtractArgsWithContext(ctx, scope, args, arg)
+		if err != nil {
+			scope.Log("xattr: %s", err.Error())
+			return
+		}
+
+		var data []byte
+		data, err = xattr.Get(arg.Filename, arg.Attribute)
```
Comment (scudette) [inline vql/darwin/xattr.go:42]: We need an accessor here too - you can get the filename with GetUnderlyingAPIFilename()

https://github.com/Velocidex/velociraptor/blob/58e92c447179c6dc7a5a09be7fe58cf0ef39c8c4/accessors/api.go#L331

File: vql/darwin/xattr.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+		if err != nil {
+			scope.Log("xattr: %s", err.Error())
+			return
+		}
+
+		output_chan <- ordereddict.NewDict().
```
Comment (scudette) [inline vql/darwin/xattr.go:48]: Maybe return a dict with all the attributes?

How large can the data be?

File: vql/darwin/xattr.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+		attributes, err := xattr.List(self.Filename)
+		if err != nil {
+			scope.Log("xattr: Failed to list attributes for filename %s: %s", self.Filename, err)
+			return nil
+		}
+
+		self.getAttributeValues(scope, attributes)
+	}
+
+	if len(self.Attributes) == 0 {
```
Comment (scudette) [inline vql/darwin/xattr.go:66]: Don't return nil from vql functions. We can return vfilter.Null or an empty dict

File: vql/darwin/xattr.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+	}
+
+	err = vql_subsystem.CheckFilesystemAccess(scope, arg.Accessor)
+	if err != nil {
+		scope.Log("xattr: %s", err)
+		return nil
+	}
+
+	self.Filename, err = accessors.GetUnderlyingAPIFilename(arg.Accessor, scope, arg.Filename)
+	self.Attributes = map[string]string{}
```
Comment (scudette) [inline vql/darwin/xattr.go:49]: Just use ordered dict here and return it directly. No need to make it a struct member

File: vql/darwin/xattr.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+)
+
+type XAttrArgs struct {
+	Filename   *accessors.OSPath `vfilter:"required,field=filename,doc=Filename to inspect."`
+	Attributes []string          `vfilter:"optional,field=attribute,doc=Attribute to collect."`
+	Accessor   string            `vfilter:"optional,field=accessor,doc=File accessor"`
+}
+
+type XAttrFunction struct {
+	Filename   string
```
Comment (scudette) [inline vql/darwin/xattr.go:25]: No need to keep struct members. It's better to not keep state in vql functions if we can help it because it's hard to know when new instances are created

File: vql/darwin/xattr.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+	return ordereddict.NewDict().
+		Set("filename", self.Filename).
+		Set("attribute", self.Attributes)
+}
+
+func (self *XAttrFunction) getAttributeValues(scope vfilter.Scope, Attributes []string) {
+	for _, attr := range Attributes {
+		value, err := xattr.Get(self.Filename, attr)
+		if err != nil {
+			scope.Log("xattr: Failled to get attribute %s from filename %s: %s", attr, self.Filename, err)
```
Comment (scudette) [inline vql/darwin/xattr.go:79]: This might generate too many logs. There is nothing much the user can do about it. Maybe drop logging? Or deduplicate the log?

File: vql/darwin/xattr_test.go [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+	"www.velocidex.com/golang/velociraptor/vql/acl_managers"
+	"www.velocidex.com/golang/velociraptor/vtesting/assert"
+)
+
+type XattrTestSuite struct {
+	test_utils.TestSuite
+}
+
+func (self *XattrTestSuite) TestXAttr() {
+	filepath := "/tmp/xattr.test"
```
Comment (scudette) [inline vql/darwin/xattr_test.go:25]: Use a temp dir to ensure the file doesn't exist already

File: vql/darwin/xattr_test.go [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+	attr1 := "vr.test"
+	attr2 := "vr.test"
+
+	_, err := os.Create(filepath)
+	if err != nil {
+		self.T().Errorf("xattr: failed to create test file: %s", err)
```
Comment (scudette) [inline vql/darwin/xattr_test.go:31]: Use assert in tests. For example assert.Error()

File: vql/darwin/xattr_test.go [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+		return
+	}
+	defer os.Remove(filepath)
+
+	xattr.Set(filepath, attr1, []byte("test-value"))
+	xattr.Set(filepath, attr2, []byte("test-value"))
+
+	ctx := context.Background()
```
Comment (scudette) [inline vql/darwin/xattr_test.go:39]: Use self.Ctx

File: vql/darwin/xattr_test.go [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+	sub_ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
```
Comment (scudette) [inline vql/darwin/xattr_test.go:40]: 100 ms seems very quick, let's give it a sec. Actually I don't think there are any cancellation points in this plugin at all so we don't really need a timeout

File: vql/darwin/xattr_test.go [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+	}
+
+	manager, err := services.GetRepositoryManager(self.ConfigObj)
+	assert.NoError(self.T(), err)
+
+	scope := manager.BuildScope(builder)
+	defer scope.Close()
+
+	testcases := []struct {
+		name string
```
Comment (scudette) [inline vql/darwin/xattr_test.go:57]: Pull these out of the function to the top of the file

File: vql/darwin/xattr_test.go [hunk 12]
```diff
@@ -80,23 +57,35 @@ func (self *XattrTestSuite) TestXAttr() {
...
-			attr: []string{"invalud.test"},
+			attr: []string{"invalid.test"},
 			pass: false,
 		},
 	}
 
+	xattr.Set(filepath, attr1, []byte("test-value"))
+	xattr.Set(filepath, attr2, []byte("test-value"))
+
+	self.Ctx = context.Background()
```
Comment (scudette) [inline vql/darwin/xattr_test.go:68]: No need to set it - it is already available

File: vql/darwin/xattr.go [hunk 13]
```diff
@@ -45,42 +42,46 @@ func (self XAttrFunction) Call(
...
 
 	return ordereddict.NewDict().
-		Set("filename", self.Filename).
-		Set("attribute", self.Attributes)
+		Set("filename", filename).
+		Set("attribute", attr)
 }
 
-func (self *XAttrFunction) getAttributeValues(scope vfilter.Scope, Attributes []string) {
+func (self *XAttrFunction) getAttributeValues(scope vfilter.Scope, Attributes []string, Filename string) map[string]string {
```
Comment (scudette) [inline vql/darwin/xattr.go:71]: just return ordereddict.Dict from here and return that into VQL

File: vql/darwin/xattr.go [hunk 14]
```diff
@@ -45,42 +42,46 @@ func (self XAttrFunction) Call(
...
-	if len(self.Attributes) == 0 {
-		return nil
+	if attr == nil {
+		return vfilter.Null{}
 	}
 
 	return ordereddict.NewDict().
-		Set("filename", self.Filename).
-		Set("attribute", self.Attributes)
+		Set("filename", filename).
```
Comment (scudette) [inline vql/darwin/xattr.go:67]: there is probably little value in echoing the same filename we just provided as an arg - so just return the attr dict directly

File: vql/darwin/xattr_test.go [hunk 15]
```diff
@@ -80,23 +57,35 @@ func (self *XattrTestSuite) TestXAttr() {
...
+	sub_ctx, cancel := context.WithCancel(self.Ctx)
```
Comment (scudette) [inline vql/darwin/xattr_test.go:69]: no need for subctx here

File: vql/darwin/xattr_test.go [hunk 16]
```diff
@@ -80,23 +57,35 @@ func (self *XattrTestSuite) TestXAttr() {
...
+	}
+
+	manager, err := services.GetRepositoryManager(self.ConfigObj)
+	assert.NoError(self.T(), err)
+
+	scope := manager.BuildScope(builder)
+	defer scope.Close()
+
 	for i, test := range testcases {
 		defer scope.Close()
```
Comment (scudette) [inline vql/darwin/xattr_test.go:86]: defer in a loop

File: vql/darwin/xattr_test.go [hunk 17]
```diff
@@ -80,23 +57,35 @@ func (self *XattrTestSuite) TestXAttr() {
...
 		ret := XAttrFunction{}.Call(sub_ctx, scope, ordereddict.NewDict().Set("filename", test.name).Set("attribute", test.attr))
```
Comment (scudette) [inline vql/darwin/xattr_test.go:87]: better line breaks - try to keep lines under 80 chars

File: vql/darwin/xattr_test.go [hunk 18]
```diff
@@ -80,23 +57,35 @@ func (self *XattrTestSuite) TestXAttr() {
...
-		if ret == nil {
-			if test.pass {
-				self.T().Errorf("xattr: test %d: Got %t, espected %t", i, false, test.pass)
-			}
-		} else {
-			if !test.pass {
-				self.T().Errorf("xattr: test %d: Got %t, espected %t", i, true, test.pass)
-			}
-		}
+		assert.Equal(self.T(), test.pass, (ret != vfilter.Null{}), "These two values should be the same. Test %d", i)
```
Comment (scudette) [inline vql/darwin/xattr_test.go:88]: maybe instead of comparing the test to null (as a bool) you can serialize the output and compare it to what is expected?

File: artifacts/testdata/files/malware.xlsm
Comment (scudette) [inline artifacts/testdata/files/malware.xlsm:1]: Can you please not remove unrelated files? Also please dont change package-lock.json as it is managed by Snyke

Comment (bmcder02) [reply artifacts/testdata/files/malware.xlsm:1]: Sorry AV removed it, and I didn't notice in the PR.
